README.md
# Zway-Astronomy

Zway Astronomy module for calculating the current solar position above the
horizon. The module provides various metrics for other automation modules
and emits events when the sun reaches certain positions.

For example, the solar altitude may be used in conjunction with the
"Motion Trigger" module ( https://github.com/maros/Zway-MotionTrigger ) to
trigger lights only if it is dark outside. Events emitted by this module
(eg. astronomy.sunrise) can be caught and processed by the "Event Watcher"
( https://github.com/maros/Zway-EventWatcher ) module to trigger custom
actions.

This module is required by the "Blind Control" module
( https://github.com/maros/Zway-BlindControl ) which manages blinds for
optimal shading.

# Configuration

## latitude,longitude

Geo coordinates for the used position

## createAzimuthDevice

If set to true, will create a device showing the current azimuth of the sun.

# Virtual Devices

This module creates a virtual device that displays the angle of the sun
above/below the horizon in degrees. Additionally the following metrics are
set and can be accessed from other automation modules.

*    metrics:azimuth: Sun position degrees
*    metrics:altitude: Altitude above horizon
*    metrics:sunrise: Top edge of the sun appears on the horizon
*    metrics:sunriseEnd: Bottom edge of the sun touches the horizon
*    metrics:goldenHourEnd: Morning golden hour ends
*    metrics:solarNoon: Sun is in the highest position
*    metrics:goldenHour: Evening golden hour starts
*    metrics:sunsetStart: Bottom edge of the sun touches the horizon
*    metrics:sunset: Sun disappears below the horizon, evening civil twilight starts
*    metrics:dusk: Evening nautical twilight starts
*    metrics:nauticalDusk: Evening astronomical twilight starts
*    metrics:night: Night starts.dark enough for astronomical observations
*    metrics:nadir: Darkest moment of the night, sun is in the lowest position
*    metrics:nightEnd: Night ends. Morning astronomical twilight starts
*    metrics:nauticalDawn: Morning nautical twilight starts)
*    metrics:dawn: Morning nautical twilight ends, morning civil twilight starts
*    metrics:trend: 'rise' or 'set'

All metrics except azimuth, altitude and trend are stored as JavaScript Date
objects, and may lay in the past since they are only calculated for the
current day. The device icon indicates day/night.

If createAzimuthDevice config is enabled, then a second virtual device
will be created.

# Events

## astronomy.sunrise

## astronomy.sunset

## astronomy.sunriseEnd

## astronomy.goldenHourEnd

## astronomy.solarNoon

## astronomy.goldenHour

## astronomy.sunsetStart

## astronomy.dusk

## astronomy.nauticalDusk

## astronomy.night

## astronomy.nadir

## astronomy.nightEnd

## astronomy.nauticalDawn

## astronomy.dawn

# Installation

The prefered way of installing this module is via the "Zwave.me App Store"
available in 2.2.0 and higher. For stable module releases no access token is
required. If you want to test the latest pre-releases use 'k1_beta' as
app store access token.

For developers and users of older Zway versions installation via git is
recommended.

```shell
cd /opt/z-way-server/automation/userModules
git clone https://github.com/maros/Zway-Astronomy.git Astronomy --branch latest
```

To update or install a specific version
```shell
cd /opt/z-way-server/automation/userModules/Astronomy
git fetch --tags
# For latest released version
git checkout tags/latest
# For a specific version
git checkout tags/1.02
# For development version
git checkout -b master --track origin/master
```

# License

Compass icon by Gregor Črešnar from the Noun Project.

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or any
later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU General Public License for more details.


CHANGES
1.10 2019 Apr 26, Fri
    * Cleanup
    * Round values to two digits
    * Update copyright

1.09 2016 Oct 26, Wed
    * Create and update devices only if configured #5
    * Fix module icon #6
    * Remove placeholder since it only confuses people
    * Add solar trend
    * Update module name

1.08 2016 Apr 20, Wed
    * Update documentation
    * Fix translation
    * Add azimut device
    * Set probeType
    * this -> self
    * Make sure we require decimal notation coordinates
    * Update installation howto

1.07 2016 Feb 26, Fri
    * Cleanup
	* Fix interval stop

1.06 2016 Feb 24, Wed
    * Update module description
    * Update to latest suncalc.js
    * Load suncalc.js with moduleBasePath prefix
    * Remove cron dependency declaration
    * Add doc link

1.05 2016 Jan 19, Tue
    * Improve description
    * Rename title/description key to make module work in appstore

1.04 2016 Jan 03, Sun
    * Use loop to handle astronomical events. Not unreliable timeouts
    * Update readme & homepage

1.03 2015 Dec 19, Sat
    * update astronomy before calling event
    * German translation

1.02 2015 Nov 15, Sun
    * Cleanups

1.01 2015 Oct 25, Sun
    * Cleanups

1.00 2015 Oct 14, Wed
    * Initial release


suncalc.js
/*
 (c) 2011-2015, Vladimir Agafonkin
 SunCalc is a JavaScript library for calculating sun/moon position and light phases.
 https://github.com/mourner/suncalc
*/

(function () { 'use strict';

// shortcuts for easier to read formulas

var PI   = Math.PI,
    sin  = Math.sin,
    cos  = Math.cos,
    tan  = Math.tan,
    asin = Math.asin,
    atan = Math.atan2,
    acos = Math.acos,
    rad  = PI / 180;

// sun calculations are based on http://aa.quae.nl/en/reken/zonpositie.html formulas


// date/time constants and conversions

var dayMs = 1000 * 60 * 60 * 24,
    J1970 = 2440588,
    J2000 = 2451545;

function toJulian(date) { return date.valueOf() / dayMs - 0.5 + J1970; }
function fromJulian(j)  { return new Date((j + 0.5 - J1970) * dayMs); }
function toDays(date)   { return toJulian(date) - J2000; }


// general calculations for position

var e = rad * 23.4397; // obliquity of the Earth

function rightAscension(l, b) { return atan(sin(l) * cos(e) - tan(b) * sin(e), cos(l)); }
function declination(l, b)    { return asin(sin(b) * cos(e) + cos(b) * sin(e) * sin(l)); }

function azimuth(H, phi, dec)  { return atan(sin(H), cos(H) * sin(phi) - tan(dec) * cos(phi)); }
function altitude(H, phi, dec) { return asin(sin(phi) * sin(dec) + cos(phi) * cos(dec) * cos(H)); }

function siderealTime(d, lw) { return rad * (280.16 + 360.9856235 * d) - lw; }


// general sun calculations

function solarMeanAnomaly(d) { return rad * (357.5291 + 0.98560028 * d); }

function eclipticLongitude(M) {

    var C = rad * (1.9148 * sin(M) + 0.02 * sin(2 * M) + 0.0003 * sin(3 * M)), // equation of center
        P = rad * 102.9372; // perihelion of the Earth

    return M + C + P + PI;
}

function sunCoords(d) {

    var M = solarMeanAnomaly(d),
        L = eclipticLongitude(M);

    return {
        dec: declination(L, 0),
        ra: rightAscension(L, 0)
    };
}


var SunCalc = {};


// calculates sun position for a given date and latitude/longitude

SunCalc.getPosition = function (date, lat, lng) {

    var lw  = rad * -lng,
        phi = rad * lat,
        d   = toDays(date),

        c  = sunCoords(d),
        H  = siderealTime(d, lw) - c.ra;

    return {
        azimuth: azimuth(H, phi, c.dec),
        altitude: altitude(H, phi, c.dec)
    };
};


// sun times configuration (angle, morning name, evening name)

var times = SunCalc.times = [
    [-0.833, 'sunrise',       'sunset'      ],
    [  -0.3, 'sunriseEnd',    'sunsetStart' ],
    [    -6, 'dawn',          'dusk'        ],
    [   -12, 'nauticalDawn',  'nauticalDusk'],
    [   -18, 'nightEnd',      'night'       ],
    [     6, 'goldenHourEnd', 'goldenHour'  ]
];

// adds a custom time to the times config

SunCalc.addTime = function (angle, riseName, setName) {
    times.push([angle, riseName, setName]);
};


// calculations for sun times

var J0 = 0.0009;

function julianCycle(d, lw) { return Math.round(d - J0 - lw / (2 * PI)); }

function approxTransit(Ht, lw, n) { return J0 + (Ht + lw) / (2 * PI) + n; }
function solarTransitJ(ds, M, L)  { return J2000 + ds + 0.0053 * sin(M) - 0.0069 * sin(2 * L); }

function hourAngle(h, phi, d) { return acos((sin(h) - sin(phi) * sin(d)) / (cos(phi) * cos(d))); }

// returns set time for the given sun altitude
function getSetJ(h, lw, phi, dec, n, M, L) {

    var w = hourAngle(h, phi, dec),
        a = approxTransit(w, lw, n);
    return solarTransitJ(a, M, L);
}


// calculates sun times for a given date and latitude/longitude

SunCalc.getTimes = function (date, lat, lng) {

    var lw = rad * -lng,
        phi = rad * lat,

        d = toDays(date),
        n = julianCycle(d, lw),
        ds = approxTransit(0, lw, n),

        M = solarMeanAnomaly(ds),
        L = eclipticLongitude(M),
        dec = declination(L, 0),

        Jnoon = solarTransitJ(ds, M, L),

        i, len, time, Jset, Jrise;


    var result = {
        solarNoon: fromJulian(Jnoon),
        nadir: fromJulian(Jnoon - 0.5)
    };

    for (i = 0, len = times.length; i < len; i += 1) {
        time = times[i];

        Jset = getSetJ(time[0] * rad, lw, phi, dec, n, M, L);
        Jrise = Jnoon - (Jset - Jnoon);

        result[time[1]] = fromJulian(Jrise);
        result[time[2]] = fromJulian(Jset);
    }

    return result;
};


// moon calculations, based on http://aa.quae.nl/en/reken/hemelpositie.html formulas

function moonCoords(d) { // geocentric ecliptic coordinates of the moon

    var L = rad * (218.316 + 13.176396 * d), // ecliptic longitude
        M = rad * (134.963 + 13.064993 * d), // mean anomaly
        F = rad * (93.272 + 13.229350 * d),  // mean distance

        l  = L + rad * 6.289 * sin(M), // longitude
        b  = rad * 5.128 * sin(F),     // latitude
        dt = 385001 - 20905 * cos(M);  // distance to the moon in km

    return {
        ra: rightAscension(l, b),
        dec: declination(l, b),
        dist: dt
    };
}

SunCalc.getMoonPosition = function (date, lat, lng) {

    var lw  = rad * -lng,
        phi = rad * lat,
        d   = toDays(date),

        c = moonCoords(d),
        H = siderealTime(d, lw) - c.ra,
        h = altitude(H, phi, c.dec);

    // altitude correction for refraction
    h = h + rad * 0.017 / tan(h + rad * 10.26 / (h + rad * 5.10));

    return {
        azimuth: azimuth(H, phi, c.dec),
        altitude: h,
        distance: c.dist
    };
};


// calculations for illumination parameters of the moon,
// based on http://idlastro.gsfc.nasa.gov/ftp/pro/astro/mphase.pro formulas and
// Chapter 48 of "Astronomical Algorithms" 2nd edition by Jean Meeus (Willmann-Bell, Richmond) 1998.

SunCalc.getMoonIllumination = function (date) {

    var d = toDays(date),
        s = sunCoords(d),
        m = moonCoords(d),

        sdist = 149598000, // distance from Earth to Sun in km

        phi = acos(sin(s.dec) * sin(m.dec) + cos(s.dec) * cos(m.dec) * cos(s.ra - m.ra)),
        inc = atan(sdist * sin(phi), m.dist - sdist * cos(phi)),
        angle = atan(cos(s.dec) * sin(s.ra - m.ra), sin(s.dec) * cos(m.dec) -
                cos(s.dec) * sin(m.dec) * cos(s.ra - m.ra));

    return {
        fraction: (1 + cos(inc)) / 2,
        phase: 0.5 + 0.5 * inc * (angle < 0 ? -1 : 1) / Math.PI,
        angle: angle
    };
};


function hoursLater(date, h) {
    return new Date(date.valueOf() + h * dayMs / 24);
}

// calculations for moon rise/set times are based on http://www.stargazing.net/kepler/moonrise.html article

SunCalc.getMoonTimes = function (date, lat, lng) {
    var t = new Date(date);
    t.setHours(0);
    t.setMinutes(0);
    t.setSeconds(0);
    t.setMilliseconds(0);

    var hc = 0.133 * rad,
        h0 = SunCalc.getMoonPosition(t, lat, lng).altitude - hc,
        h1, h2, rise, set, a, b, xe, ye, d, roots, x1, x2, dx;

    // go in 2-hour chunks, each time seeing if a 3-point quadratic curve crosses zero (which means rise or set)
    for (var i = 1; i <= 24; i += 2) {
        h1 = SunCalc.getMoonPosition(hoursLater(t, i), lat, lng).altitude - hc;
        h2 = SunCalc.getMoonPosition(hoursLater(t, i + 1), lat, lng).altitude - hc;

        a = (h0 + h2) / 2 - h1;
        b = (h2 - h0) / 2;
        xe = -b / (2 * a);
        ye = (a * xe + b) * xe + h1;
        d = b * b - 4 * a * h1;
        roots = 0;

        if (d >= 0) {
            dx = Math.sqrt(d) / (Math.abs(a) * 2);
            x1 = xe - dx;
            x2 = xe + dx;
            if (Math.abs(x1) <= 1) roots++;
            if (Math.abs(x2) <= 1) roots++;
            if (x1 < -1) x1 = x2;
        }

        if (roots === 1) {
            if (h0 < 0) rise = i + x1;
            else set = i + x1;

        } else if (roots === 2) {
            rise = i + (ye < 0 ? x2 : x1);
            set = i + (ye < 0 ? x1 : x2);
        }

        if (rise && set) break;

        h0 = h2;
    }

    var result = {};

    if (rise) result.rise = hoursLater(t, rise);
    if (set) result.set = hoursLater(t, set);

    if (!rise && !set) result[ye > 0 ? 'alwaysUp' : 'alwaysDown'] = true;

    return result;
};


// export as AMD module / Node module / browser variable
if (typeof define === 'function' && define.amd) define(SunCalc);
else if (typeof module !== 'undefined') module.exports = SunCalc;
else window.SunCalc = SunCalc;

}());

index.js
/*** Astronomy Z-Way HA module *******************************************

Version: 1.10
(c) Maroš Kollár, 2015-2017
-----------------------------------------------------------------------------
Author: Maroš Kollár <maros@k-1.com>
Description:
    This module checks weather updates via weatherundergound.com

******************************************************************************/

function Astronomy (id, controller) {
    // Call superconstructor first (AutomationModule)
    Astronomy.super_.call(this, id, controller);

    this.latitude       = undefined;
    this.longitude      = undefined;
    this.vDev           = {};
    this.timer          = undefined;
}

inherits(Astronomy, AutomationModule);

_module = Astronomy;

// Helper function
function round(number) {
    var factor = Math.pow(10, 2);
    return Math.round(number * factor) / factor;
}

// ----------------------------------------------------------------------------
// --- Module instance initialized
// ----------------------------------------------------------------------------

Astronomy.prototype.events = [
    'sunrise',          // sunrise (top edge of the sun appears on the horizon)
    'sunriseEnd',       // sunrise ends (bottom edge of the sun touches the horizon)
    'goldenHourEnd',    // morning golden hour (soft light, best time for photography) ends
    'solarNoon',        // solar noon (sun is in the highest position)
    'goldenHour',       // evening golden hour starts
    'sunsetStart',      // sunset starts (bottom edge of the sun touches the horizon)
    'sunset',           // sunset (sun disappears below the horizon, evening civil twilight starts)
    'dusk',             // dusk (evening nautical twilight starts)
    'nauticalDusk',     // nautical dusk (evening astronomical twilight starts)
    'night',            // night starts (dark enough for astronomical observations)
    'nadir',            // nadir (darkest moment of the night, sun is in the lowest position)
    'nightEnd',         // night ends (morning astronomical twilight starts)
    'nauticalDawn',     // nautical dawn (morning nautical twilight starts)
    'dawn'              // dawn (morning nautical twilight ends, morning civil twilight starts)
];

Astronomy.prototype.init = function (config) {
    Astronomy.super_.prototype.init.call(this, config);

    var self = this;

    // See https://github.com/mourner/suncalc
    executeFile(self.moduleBasePath()+"/suncalc.js");

    self.latitude       = config.latitude.toString();
    self.longitude      = config.longitude.toString();
    config.createAltitudeDevice = true;

    var langFile        = self.controller.loadModuleLang("Astronomy");
    _.each(self.events,function(event) {
        self[event+'Timer'] = undefined;
    });

    _.each(['altitude','azimuth'],function(type) {
        var configKey = 'create' + type.charAt(0).toUpperCase() + type.substring(1) + 'Device';
        var configDevice = config[configKey];
        if (configDevice === true
            || _.isUndefined(configDevice)) {
            self.vDev[type]= self.controller.devices.create({
                deviceId: "Astronomy_"+self.id+"_"+type,
                defaults: {
                    deviceType: "sensorMultilevel",
                    metrics: {
                        icon: 'icon.png',
                        title: langFile[type+'_device']
                    }
                },
                overlay: {
                    probeType: 'astronomy_sun_'+type,
                    metrics: {
                        scaleTitle: "°"
                    }
                },
                moduleId: self.id
            });
        }
    });

    self.interval = setInterval(function() {
        self.updateCalculation(self);
    }, 60*1000);

    self.updateCalculation();
};

Astronomy.prototype.stop = function () {
    var self = this;

    _.each(['altitude','azimuth'],function(type) {
        if (typeof(self.vDev[type]) !== 'undefined') {
            self.controller.devices.remove(self.vDev[type].id);
            self.vDev[type] = undefined;
        }
    });

    clearInterval(self.interval);
    self.timer = undefined;

    Astronomy.super_.prototype.stop.call(self);
};

// ----------------------------------------------------------------------------
// --- Module methods
// ----------------------------------------------------------------------------

Astronomy.prototype.updateCalculation = function () {
    var self        = this;
    //var langFile    = self.controller.loadModuleLang("Astronomy");
    var now         = new Date();
    var position    = SunCalc.getPosition(now, self.config.latitude, self.config.longitude);
    var times       = SunCalc.getTimes(now, self.config.latitude, self.config.longitude);
    var azimuth     = round(position.azimuth * 180 / Math.PI + 180);
    var altitude    = round(position.altitude * 180 / Math.PI);
    var previous    = parseFloat(self.vDev.altitude.get("metrics:level") || altitude);
    var mode;

    console.log("[Astronomy] Calculate");
    if (altitude < -2) {
        mode = 'night';
    } else {
        mode = 'day';
    }

    if (! _.isUndefined(self.vDev.altitude)) {
        self.vDev.altitude.set("metrics:icon", "/ZAutomation/api/v1/load/modulemedia/Astronomy/altitude_"+mode+".png");
        self.vDev.altitude.set("metrics:level",altitude);
        self.vDev.altitude.set("metrics:azimuth",azimuth);
        self.vDev.altitude.set("metrics:altitude",altitude);
        self.vDev.altitude.set("metrics:trend",(previous <= altitude) ? 'rise':'set');
    }

    if (! _.isUndefined(self.vDev.azimuth)) {
        self.vDev.azimuth.set("metrics:icon", "/ZAutomation/api/v1/load/modulemedia/Astronomy/azimuth_"+mode+".png");
        self.vDev.azimuth.set("metrics:level",azimuth);
        self.vDev.azimuth.set("metrics:azimuth",azimuth);
        self.vDev.azimuth.set("metrics:altitude",altitude);
    }

    _.each(self.events,function(event) {
        if (! _.isUndefined(self.vDev.altitude)) {
            self.vDev.altitude.set("metrics:"+event,times[event]);
        }
        if (times[event].getHours() === now.getHours()
            && times[event].getMinutes() === now.getMinutes()
            && times[event].getDate() === now.getDate()) {
            console.log("[Astronomy] Event "+event);
            self.controller.emit("astronomy."+event);
        }
    });
};
